File: Java/1-Two Sum.java
//1. Two Sum

//Brute Force O(n^2)
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int[] ans = new int[2];
        for(int i = 0; i < nums.length; i ++){
            for(int j = nums.length - 1; j > i; j --){
                if(target == (nums[i] + nums[j])){
                    ans[0] = i ;
                    ans[1] = j;
                    return ans;
                }
            }
        }
        return null;
        
    }
}


//HashMap O(n)
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int[] ans = new int[2];
        HashMap<Integer, Integer> seen = new HashMap<>();
        for(int i = 0 ; i < nums.length; i++){
            if(seen.containsKey(target - nums[i])){
                ans[0] = seen.get(target - nums[i]);
                ans[1] = i;
                return ans;
                
            }
            seen.put(nums[i], i);
            
        }
     return ans;   
    }
}

File: Python3/1. Two Sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        i,j = 0,1
        ans = []
        while(i < len(nums)):
            j = j + i
            while(j < len(nums)):
                    if(nums[i] + nums[j] == target):
                        ans.append(i)
                        ans.append(j)
                        return ans

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        
        checkMap = {} # diff, index

        for i in range(0, len(nums)):
            diff = target - nums[i]
            if diff in checkMap:
                return [i , checkMap[diff]]

            checkMap[nums[i]] = i


##NOTE
#- output: indices of tow numbers
#- input: array of integers
#- constraints: ip has one soln, no repition, return answere in any order, atleast 2 num in array
#-Soln:
#	- 2 ptr
			#- T:O(n^2), S:O(1)
			#- drawback: T:O(n^2)
			#- check: 
	#- Data Structure : HasMap
		#	- T:O(n), S:O(n)
		#	- drawback: S:O(n)
		#	- check: 
